packages/drivers/driver-base/src/documentDeltaConnection.ts

```diff
@@ -305,16 +305,40 @@ export class DocumentDeltaConnection
         this.submitManager.add("submitSignal", [message]);
     }
 
+    /**
+     * Disconnect from the websocket, and permanently disable this DocumentDeltaConnection.
+     */
+    public close() {
+        // close() assumes silent close - no events fired to avoid reentrancy
+        this.removeAllListeners();
+        this.disconnect(false, "client closing connection");
+    }
+
     /**
      * Disconnect from the websocket, and permanently disable this DocumentDeltaConnection.  Subclasses which
      * override this method must set the "closed" flag after disconnecting.
      * @param socketProtocolError - true if error happened on socket / socket.io protocol level
      *  (not on Fluid protocol level)
+     * @param reason - reason for disconnect
      */
-    public disconnect(socketProtocolError: boolean = false) {
+    protected disconnect(socketProtocolError: boolean, reason: string) {
+        // We set the closed flag as a part of the contract for overriding the disconnect method. This is used by
+        // DocumentDeltaConnection to determine if emitting on the socket is allowed, which is important since
+        // OdspDocumentDeltaConnection reuses the socket rather than truly disconnecting it.  Note that below we may
+        // still send disconnect_document which is allowed; this is only intended to prevent normal messages from
+        // being emitted.
+        if (this.closed) {
+            return;
+        }
+        this.closed = true;
+
         this.removeTrackedListeners(false);
+        this.disconnectCore(socketProtocolError, reason);
+        this.removeAllListeners();
```

Comment:
I'm not a fan of removing event listeners in this manner in general, it leaves the listeners unaware that their listener has been modified which can be hard to account for in internal state.

Reply:
Is your comment applicable only to this function, or in general? I believe close() method has to unregister, otherwise this memory is leaked - that's standard flow we use in other places on closure.

This particular case is a bit weird, because I do not understand when disconnect() can be called not through close() (using function names of this PR), when there are actually callbacks installed by DeltaManager. So to some extent, that's safety mechanism. Also unregistering listeners should follow removeTrackedListeners(false), as no events are going to be fired anyway (we unsubscribed from all events on socket).

Reply:
In general.  My preference is for listeners to always unregister themselves when it's appropriate, rather than the emitter doing the unregister.

Reply:
Take a look - I'm not sure latest iteration is better, but now the caller calls removeAllListeners()